pytorch_pfn_extras/reporting.py

```diff
@@ -303,10 +303,9 @@ def state_dict(self):
         return state
 
     def load_state_dict(self, to_load):
-        self._x = to_load['_x']
-        self._x2 = to_load['_x2']
-        self._n = to_load['_n']
-
+        self._x = numpy.float64(to_load['_x'])
```

Comment:
I think `numpy.asarray` is better to preserve dtype.
Note that values inside `to_load` maybe any of Python float (`1.0`), NumPy array or torch.Tensor...

Reply:
I tried to replace `numpy.float64` with `numpy.asarray`, but I got `TypeError: can't convert CUDA tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.`.


So, I prepared a type aware converter method like follows.

```Python
def load_as_numpy_array(x):
    # return numpy.asarray(x)
    if isinstance(x, torch.Tensor):
        return torch.detach(x).cpu()#.numpy()
    else:
        return numpy.asarray(x)
```


```Python
    def load_state_dict(self, to_load):
        self._x = load_as_numpy_array(to_load['_x'])
        self._x2 = load_as_numpy_array(to_load['_x2'])
        self._n = load_as_numpy_array(to_load['_n'])
```

But still `TypeError` occurs for some metrics.

```
  File "./lib/python3.8/site-packages/pytorch_pfn_extras/training/extensions/log_report.py", line 90, in __call__
    summary.add(observation)
  File "./lib/python3.8/site-packages/pytorch_pfn_extras/reporting.py", line 350, in add
    summaries[k].add(v, weight=w)
  File "./lib/python3.8/site-packages/pytorch_pfn_extras/reporting.py", line 272, in add
    self._x += weight * value
TypeError: unsupported operand type(s) for +=: 'numpy.ndarray' and 'Tensor'
```

Here is a reproducer.

```Python
    summary = ppe.reporting.Summary()
    summary.add(torch.tensor([[15.0]], requires_grad=True))

    f = io.BytesIO()
    torch.save(summary.state_dict(), f)
    summary.load_state_dict(torch.load(io.BytesIO(f.getvalue())))

    summary.add(torch.tensor(5.0)) # torch.tensor input
    mean = summary.compute_mean() # => TypeError: unsupported operand type(s) for +=: 'numpy.ndarray' and 'Tensor'
    numpy.testing.assert_allclose(mean, 10)
```

Finally this patch solves this issue.

```Python
    def add(self, value, weight=1):
        """Adds a scalar value.

        Args:
            value: Scalar value to accumulate. It is either a NumPy scalar or
                a zero-dimensional array (on CPU or GPU).
            weight: An optional weight for the value. It is a NumPy scalar or
                a zero-dimensional array (on CPU or GPU).
                Default is 1 (integer).

        """
        value = load_as_numpy_array(value) #### <- patched
        self._x += weight * value
        self._x2 += weight * value * value
        self._n += weight
```

I feel these patches are too naive, so I would like to find additional ideas to make this simple.

Thank you.

Reply:
Sorry for the delayed reply  :bow: I will fix and and add tests in #60.